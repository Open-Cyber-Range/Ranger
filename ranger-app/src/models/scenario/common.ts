export type Source = {
  name: string;
  version: string;
};
