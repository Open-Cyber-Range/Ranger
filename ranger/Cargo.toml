[package]
name = "ranger"
version = "2.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[package.metadata.deb]
name = "ranger-server"
maintainer = "C14 developers <developers@cr14.ee>"
copyright = "MIT"
depends = "$auto, systemd"
section = "utility"
priority = "optional"
changelog = "changelog"
assets = [
  [
    "target/release/ranger",
    "/var/opt/ranger/bin/",
    "755",
  ],
  [
    "assets/default-production-config.yml",
    "/etc/opt/ranger/ranger/config.yml",
    "644",
  ],
  [
    "../ranger-app/build/assets/*",
    "/etc/opt/ranger/ranger-app/assets",
    "644",
  ],
  [
    "../ranger-app/build/*",
    "/etc/opt/ranger/ranger-app/",
    "644",
  ],
]
systemd-units = { unit-name = "ranger", unit-scripts = "assets/unit-scripts" }

[dependencies]
actix = "0.13"
actix-multipart = "0.6"
actix-web = "4"
actix-web-actors = "4"
actix-web-grants = "3"
actix-http = { version = "3", features = ["http2", "ws"] }
actix-web-httpauth = "0.8"
anyhow = "1"
async-trait = "0.1"
bigdecimal = { version = "0.4", features = ["serde"] }
chrono = { version = "0.4", features = ["serde"] }
diesel = { version = "2.1", features = ["mysql", "r2d2", "chrono", "numeric"] }
diesel_migrations = "2.1"
env_logger = "0.9"
futures = "0.3"
futures-util = "0.3"
jsonwebtoken = "8"
keycloak = { git = "https://github.com/ebakoba/keycloak", branch = "master" }
log = "0.4"
lazy_static = "1.4"
lettre = { version = "0.10", default-features = false, features = [
  "smtp-transport",
  "pool",
  "hostname",
  "builder",
  "rustls",
  "rustls-tls",
] }
sha3 = "0.10"
digest = "0.10"
rand = "0.8"
ranger-grpc = { path = "../ranger-grpc" }
reqwest = { version = "0.11", default-features = false, features = [
  "json",
  "rustls",
  "rustls-tls",
  "tokio-rustls",
  "hyper-rustls",
  "rustls-native-certs",
] }
sdl-parser = "0"
serde = { version = "1.0", features = ["derive"] }
serde_json = "1"
serde_yaml = "0.9"
semver = "1"
tempfile = "3"
thiserror = "1"
toml-query = "0.10"
toml = "0.5.6"
tokio = { version = "1", features = [
  "rt-multi-thread",
  "time",
  "fs",
  "macros",
  "net",
  "rt",
] }
tonic = "0.8"
uuid = { version = "1.1", features = ["v4", "serde", "fast-rng"] }

[dev-dependencies]
actix-rt = "2"
insta = { version = "1", features = ["redactions", "yaml"] }
sdl-parser = { version = "0", features = ["test"] }
