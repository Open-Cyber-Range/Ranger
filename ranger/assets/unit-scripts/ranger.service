[Unit]
Description=ranger
After=network.target

[Service]
ExecStart=/var/opt/ranger/bin/ranger /etc/opt/ranger/ranger/config.yml
Type=simple
Restart=on-failure
RestartSec=10
StartLimitInterval=10m
StartLimitBurst=5

[Install]
WantedBy=multi-user.target
