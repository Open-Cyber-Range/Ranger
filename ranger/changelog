ranger (0.2.0) stretch; urgency=medium

  * Initial Release
  * Adds web application support

 -- Kaarel Allemann <kaarel.allemann@cr14.ee>  Sat, 05 Nov 2022 19:23:51 +0200
