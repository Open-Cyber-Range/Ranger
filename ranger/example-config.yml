host: localhost
port: 8080
repository:
  folder: /tmp/test-repo
  username: some-username
  email: some@email.com
package_folder: /tmp/packages
