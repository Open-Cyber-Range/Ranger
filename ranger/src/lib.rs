pub(crate) mod configuration;
mod constants;
pub mod errors;
pub mod models;
pub mod routes;
pub(crate) mod schema;
pub(crate) mod services;
pub(crate) mod utilities;

use crate::services::database::Database;
use actix::{Actor, Addr};
use anyhow::Result;
use configuration::{read_configuration, Configuration};
use log::{error, info};
use services::{
    deployer::{DeployerDistribution, DeployerFactory},
    deployment::DeploymentManager,
    mailer::Mailer,
    scheduler::Scheduler,
    websocket::WebSocketManager,
};

pub struct AppState {
    pub database_address: Addr<Database>,
    pub deployment_manager_address: Addr<DeploymentManager>,
    pub configuration: Configuration,
    pub websocket_manager_address: Addr<WebSocketManager>,
}

impl AppState {
    pub fn new(
        configuration: &configuration::Configuration,
        distributor: &Addr<DeployerDistribution>,
        database: &Addr<Database>,
        websocket_manager: &Addr<WebSocketManager>,
    ) -> Self {
        let schduler_address = Scheduler::new().start();
        AppState {
            database_address: database.clone(),
            deployment_manager_address: DeploymentManager::new(
                schduler_address,
                distributor.clone(),
                database.clone(),
                configuration.deployment_groups.clone(),
                configuration.default_deployment_group.to_string(),
            )
            .start(),
            configuration: configuration.to_owned(),
            websocket_manager_address: websocket_manager.clone(),
        }
    }
}

pub async fn app_setup(environment_arguments: Vec<String>) -> Result<(String, u16, AppState)> {
    let configuration = read_configuration(environment_arguments)?;
    let deployer_factory = DeployerFactory::new(&configuration.deployers)
        .await
        .unwrap_or_else(|error| panic!("Failed to create deployer distribution: {error}"))
        .start();

    let deployer_distributor = DeployerDistribution::new(
        deployer_factory,
        configuration.deployers.keys().cloned().collect(),
    )
    .await
    .unwrap_or_else(|error| panic!("Failed to create deployer distribution: {error}"))
    .start();

    let websocket_manager = WebSocketManager::new().start();
    let database = Database::try_new(&configuration.database_url, &websocket_manager)
        .unwrap_or_else(|error| {
            panic!(
                "Failed to create database connection to {} due to: {error}",
                &configuration.database_url
            )
        })
        .start();

    send_mail(configuration.clone());

    let app_state = AppState::new(
        &configuration,
        &deployer_distributor,
        &database,
        &websocket_manager,
    );
    Ok((configuration.host, configuration.port, app_state))
}

pub fn send_mail(configuration: Configuration) {
    if let Some(mailer_configuration) = configuration.mailer_configuration {
        match Mailer::send_mail(
            mailer_configuration,
            "real.person@real-email.com".to_string(),
        ) {
            Ok(_) => info!("Mail sent successfully!"),
            Err(e) => error!("Mailer failed: {:?}", e),
        }
    }
}
