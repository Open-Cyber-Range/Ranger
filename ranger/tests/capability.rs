mod common;

#[cfg(test)]
mod tests {
    use crate::common::create_mock_capability_server;
    use actix::Actor;
    use actix_rt::System;
    use anyhow::{Error, Result};
    use ranger::{capability::GetCapabilities, node::NodeClient};
    use ranger_grpc::Capabilities;

    #[test]
    fn successful_capability_response() -> Result<()> {
        let socket_address = create_mock_capability_server().run()?;
        let system = System::new();
        let capabilities = system.block_on(async {
            let node_deployer_client = NodeClient::new(format!("http://{}", socket_address))
                .await?
                .start();

            let capabilities = node_deployer_client.send(GetCapabilities).await??;
            Ok::<Capabilities, Error>(capabilities)
        })?;

        insta::assert_debug_snapshot!(capabilities);
        Ok(())
    }
}
