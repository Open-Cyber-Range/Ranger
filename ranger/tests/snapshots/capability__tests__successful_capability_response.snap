---
source: ranger/tests/capability.rs
expression: capabilities
---
Capabilities {
    values: [
        VirtualMachine,
    ],
}
